### DISCUSSION TITLE
Help with migrating skills from Mycroft to OVOS

### DISCUSSION THREAD

**Discussion by erikjber** [OpenVoiceOS/ovos-buildroot#168] [Q&A] [answered]
I have a few skills I wrote for Mycroft (stuff that helps my in my day-to-day) that I want to migrate to OVOS now that Mycroft is winding down.  

I've tried the following:

- `osm install git+<git repo of skill>`
- `git clone <git repo of skill` while in the local skill folder (`/home/ovos/.local/share/mycroft/skills`). 
-  Doing the above an adding the name of the skill to the file `/home/ovos/.local/share/mycroft/skills.json`
- Copying the skill folder to `/home/ovos/.config/mycroft/skills`.  

The result is always the same: The skill is not shown in `mana get-skills-list` or if I open `ovos-cli-client` and type `:skills`.  

I've tried everything, including reboots after every change.  The file permissions look OK, everything is owned by `ovos:ovos`. 

I have activated local skills with `osm enable local`.

Do I need to add the skills to some config file somewhere?  Do I need to port them to make the compatible with OVOS? Am I using the wrong folder?

**Comment by NeonDaniel** [comment]
>Copying the skill folder to /home/ovos/.config/mycroft/skills

This should be `.local/share/mycroft/skills` rather than `.config/mycroft/skills` but the osm installation should have taken care of it.

Do you see any errors in the skills log (`grep -i error ~/.local/state/mycroft/skills.log`) that might indicate why the skill isn't loading?

**Reply by erikjber** [reply]
> This should be .local/share/mycroft/skills rather than .config/mycroft/skills but the osm installation should have taken care of it.  

Yes, I tried both.  


> Do you see any errors in the skills log (grep -i error ~/.local/state/mycroft/skills.log) that might indicate why the skill isn't loading? 

No, nothing. There are a few error messages related to other skills, but nothing about mine. 

> 2023-04-29 17:29:20.120 - skill-ovos-wikipedia.openvoiceos - ERROR - Failed to stop skill: skill-ovos-wikipedia.openvoiceos
>     raise RuntimeError("bus not set, did you call self.bind() ?")
> RuntimeError: bus not set, did you call self.bind() ?
> 2023-04-29 17:29:21.036 - skill-alerts.neongeckocom - ERROR - Skill specific shutdown function encountered an error: bus not set, did you call self.bind() ?
> 2023-04-29 17:29:21.107 - skill-ovos-weather.openvoiceos - ERROR - Failed to stop skill: skill-ovos-weather.openvoiceos
>     raise RuntimeError("bus not set, did you call self.bind() ?")
> RuntimeError: bus not set, did you call self.bind() ?
> 2023-04-29 17:29:35.401 - skills - skill_caffeinewiz:_get_new_info:473 - ERROR - An error occurred during the CaffeineWiz update: 'state'
> 2023-04-29 17:29:37.580 - skills - ovos_workshop.resource_files:_locate:252 - ERROR - Could not find resource file HowAreYou.intent
> 2023-04-29 17:29:37.581 - skill-ovos-hello-world.openvoiceos - ERROR - Unable to find "HowAreYou.intent"
> 2023-04-29 17:30:54.796 - skills - skill_caffeinewiz:_get_new_info:473 - ERROR - An error occurred during the CaffeineWiz update: 'state'
> 2023-04-29 17:30:58.558 - skills - ovos_workshop.resource_files:_locate:252 - ERROR - Could not find resource file HowAreYou.intent
> 2023-04-29 17:30:58.558 - skill-ovos-hello-world.openvoiceos - ERROR - Unable to find "HowAreYou.intent"

**Comment by erikjber** [comment]
This is the output when I install a new skill using osm:  

> $ osm install git+_github url_
> 2023-04-30 17:06:33.587 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /usr/local/lib/python3.9/dist-packages/mycroft/configuration/mycroft.conf loaded
> 2023-04-30 17:06:33.781 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /etc/mycroft/mycroft.conf loaded
> 2023-04-30 17:06:34.000 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /home/ovos/.config/mycroft/web_cache.json loaded
> 2023-04-30 17:06:34.228 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /home/ovos/.config/mycroft/mycroft.conf loaded
> 2023-04-30 17:06:34.444 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/etc/xdg/mycroft/mycroft.conf' not defined, skipping
> 2023-04-30 17:06:34.630 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/home/ovos/.mycroft/mycroft.conf' not defined, skipping
> git+_github url_ does not look like a valid skill url, do you want to enable search? [y/N]: y
> 2023-04-30 17:06:37.310 - OVOS - ovos_utils.skills:get_skills_folder:93 - WARNING - This reference is deprecated, use `ovos_utils.skills.locations.get_default_skill_dir
> 2023-04-30 17:06:37.332 - OVOS - ovos_skills_manager.appstores:bootstrap:51 - INFO - Bootstrapping InstalledSkills, this might take a while!
> 2023-04-30 17:06:37.335 - OVOS - ovos_skills_manager.appstores:bootstrap:58 - DEBUG - Downloading skill list
> 2023-04-30 17:06:37.337 - OVOS - ovos_utils.skills:get_skills_folder:93 - WARNING - This reference is deprecated, use `ovos_utils.skills.locations.get_default_skill_dir
> 2023-04-30 17:06:37.367 - OVOS - ovos_skills_manager.appstores:sync_skills_list:77 - INFO - Synced skill: _github url_
> 2023-04-30 17:06:37.370 - OVOS - ovos_skills_manager.appstores:sync_skills_list:89 - DEBUG - Adding new skill to db
> NO RESULTS
>

**Comment by JarbasAl** [comment]
`osm install --skill url` without `git+`

osm is interactive, you can just type `osm install` and follow the prompts

if the skill has a setup.py the recommended way is to `pip install git+url`, most skills are also published in pypi and can be installed via `pip install ovos-skill-xxx`

OSM is more off an interface to marketplaces/skill stores, it is not a core component or needed for the most part, and in fact is lagging behind

**Reply by erikjber** [reply]
`osm install url` did not work either. Do I need to add a `setup.py` file to my skills?

**Reply by JarbasAl** [reply]
apologies  `osm install --skill url`

also note that osm is interactive, you can just type `osm install` and follow the prompts

```
$ osm install
2023-04-30 18:49:15.608 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /home/miro/.venvs/skill-ovos-fallback-unknown/lib/python3.9/site-packages/mycroft/configuration/mycroft.conf loaded
2023-04-30 18:49:15.639 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/etc/mycroft/mycroft.conf' not defined, skipping
2023-04-30 18:49:15.685 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/home/miro/.config/mycroft/web_cache.json' not defined, skipping
2023-04-30 18:49:15.720 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/home/miro/.config/mycroft/mycroft.conf' not defined, skipping
2023-04-30 18:49:15.748 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/etc/xdg/mycroft/mycroft.conf' not defined, skipping
2023-04-30 18:49:15.775 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/home/miro/.config/kdedefaults/mycroft/mycroft.conf' not defined, skipping
2023-04-30 18:49:15.801 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/home/miro/.mycroft/mycroft.conf' not defined, skipping
Skill to install (url or search term): 

```

**Reply by erikjber** [reply]
> $  osm install --skill _skill url_
> 2023-04-30 18:45:32.294 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /usr/local/lib/python3.9/dist-packages/mycroft/configuration/mycroft.conf loaded
> 2023-04-30 18:45:32.478 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /etc/mycroft/mycroft.conf loaded
> 2023-04-30 18:45:32.680 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /home/ovos/.config/mycroft/web_cache.json loaded
> 2023-04-30 18:45:32.874 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /home/ovos/.config/mycroft/mycroft.conf loaded
> 2023-04-30 18:45:33.076 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/etc/xdg/mycroft/mycroft.conf' not defined, skipping
> 2023-04-30 18:45:33.252 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/home/ovos/.mycroft/mycroft.conf' not defined, skipping
> Usage: osm install [OPTIONS] [SKILL]
> Try 'osm install --help' for help.
> 
> Error: No such option: --skill
> 

Using the interactive option `osm install` and follwoing the prompts results in this:

> $ osm install
> 2023-04-30 18:49:11.771 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /usr/local/lib/python3.9/dist-packages/mycroft/configuration/mycroft.conf loaded
> 2023-04-30 18:49:11.979 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /etc/mycroft/mycroft.conf loaded
> 2023-04-30 18:49:12.170 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /home/ovos/.config/mycroft/web_cache.json loaded
> 2023-04-30 18:49:12.386 - OVOS - ovos_config.models:load_local:105 - DEBUG - Configuration /home/ovos/.config/mycroft/mycroft.conf loaded
> 2023-04-30 18:49:12.590 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/etc/xdg/mycroft/mycroft.conf' not defined, skipping
> 2023-04-30 18:49:12.774 - OVOS - ovos_config.models:load_local:111 - DEBUG - Configuration '/home/ovos/.mycroft/mycroft.conf' not defined, skipping
> Skill to install (url or search term): _skill url_
> 2023-04-30 18:49:22.211 - OVOS - ovos_skills_manager.github:get_skill_data:39 - DEBUG - default branch extracted from github api: master
> 2023-04-30 18:49:31.140 - OVOS - ovos_skills_manager.github.raw:get_manifest_from_github_url:333 - WARNING - _skill url_/master/manifest.yml contains an invalid manifest, attempting recovery
> 2023-04-30 18:49:31.142 - OVOS - ovos_skills_manager.github:get_requirements_json:209 - DEBUG - No manifest file available
> 2023-04-30 18:49:31.785 - OVOS - ovos_skills_manager.github:get_requirements_json:219 - DEBUG - No python requirements file available
> 2023-04-30 18:49:32.084 - OVOS - ovos_skills_manager.github:get_requirements_json:226 - DEBUG - No skill requirements file available
> 
> Search Results:
>     appstore - branch - url 
> 1 - unknown - master _skill url_
> 0 - cancel installation
> 
> Select an option: 1
> Do you want to install master _skill url_ ? [y/N]: y
> 2023-04-30 18:49:48.540 - OVOS - ovos_utils.skills:get_skills_folder:93 - WARNING - This reference is deprecated, use `ovos_utils.skills.locations.get_default_skill_dir
> 2023-04-30 18:49:48.553 - OVOS - ovos_skills_manager.skill_entry:install:327 - INFO - Installing skill: _skill url_ from branch: master
> 2023-04-30 18:49:48.555 - OVOS - ovos_skills_manager.skill_entry:install:347 - INFO - Downloading _skill url_   

Afterwards the git repo is checked out in `~/.local/share/mycroft/skills/`, but it still does not work or even show up in the list of skills.

**Comment by builderjer** [comment]
Do you have a link to your skill you are trying to install? Or would you rather not?  It might help in debugging why it will not load

**Reply by erikjber** [reply]
Sure thing! These are the three I have tested so far:    
https://github.com/erikjber/simple-shopping-list-skill   
https://github.com/erikjber/week-number-skill  
https://github.com/erikjber/tempereature-reader-skill

**Comment by erikjber** [comment]
Progress!  

If I run `ovos-skill-launcher week-numer-skill /home/ovos/.local/share/mycroft/skills/week-number-skill.erikjber` the skill starts working. 

It still won't show up if run `:skills` in `ovos-cli-client`, but it works. Is there some way I can have OVOS do this automatically? 

It works if I put it in my `crontab` and reboot, but I suspect that's not the recommended way of doing things.

**Reply by JarbasAl** [reply]
from the other thread for discoverability
__________________
that command is meant to launch skills standalone, for example in their own docker container or just for easy development. It's a new feature we just introduced

:skills command has not been updated in a long time, ovos-cli-client is in maintenance mode and not getting active development, we just fix bugs as they are reported

we are diagnosing issues around the skill loading from folder, we got some more similar reports in our chat recently and are looking into it

**Comment by madgrizzle** [comment, marked as answer]
@erikjber, I had same issue and as a work around, I added /home/xxx/.local/share/mycroft/skills as an extra directory in the mycroft.config file.  It's not checking that directory on my install for whatever reason and this forces it to.

  ```
// General skill values
  "skills": {

    // don't start loading skills until internet is detected
    // this config value is not present in mycroft-core (internet is required)
    // ovos-core expects that some instances will be running fully offline
    // DEPRECATED: specify skill loading requirements in individual skills instead
    "wait_for_internet": false,

    // relative to "data_dir"
    "directory": "skills",
    "extra_directories": ["/home/john/.local/share/mycroft/skills"],
```

**Reply by erikjber** [reply]
That works! I added the `extra_directories` line to `/usr/local/lib/python3.9/dist-packages/ovos_config/mycroft.conf` and rebooted, and now my skills are active and working. They even show up when I run `:skills` in `ovos-cli-client`. Thanks, @madgrizzle !

**Reply by JarbasAl** [reply]
i believe this issue has been fixed by https://github.com/OpenVoiceOS/ovos-core/pull/323 and the workaround is no longer needed

**Comment by Bulwagga** [comment]
This worked for me as well.  Any way to get this answer more visibility? This was a hard answer to find. :) Thanks

**Comment by builderjer** [comment]
I will put this in a new build, until this is fixex upstream. I will try a new image with a few issues addressed  this weekend

**Comment by Bulwagga** [comment]
Much appreciated.  I'll download and test when I see it posted.

**Comment by mikejgray** [comment]
For future readers, there is a helper project to automatically make some of the conversions necessary. OSM is somewhat on its last legs and will probably be converted to a wrapper over a `pip install` in the future.

https://github.com/mikejgray/ovos-skill-projen#retrofitting-mycroft-skills-wip

If you prefer not using the script or it doesn't work the way you expect (please do file an issue), there's a list of steps there to do the conversion manually.

Happy retrofitting!
